src/utils/api.js

```diff
@@ -31,6 +33,98 @@ export const formatActions = (actionsByDayObject) => {
   });
 };
 
+// Takes array of action objects and returns an object with {key: value} pairs of {hourOfDay: #-of-actions}
+export const getActionsByTimeOfDay = (actions) => {
+  const actionsByTimeOfDay = {
+    'Late night': 0,
+    'Early morning': 0,
+    Morning: 0,
+    Afternoon: 0,
+    Evening: 0,
+    Night: 0,
+  };
+  actions.forEach((action) => {
+    const actionHourOfDay = action.createdAt.slice(11, 13);
+    if (actionHourOfDay >= 0 && actionHourOfDay <= 4) {
+      actionsByTimeOfDay['Late night'] += 1;
+    } else if (actionHourOfDay > 4 && actionHourOfDay <= 8) {
+      actionsByTimeOfDay['Early morning'] += 1;
+    } else if (actionHourOfDay > 8 && actionHourOfDay <= 12) {
+      actionsByTimeOfDay.Morning += 1;
+    } else if (actionHourOfDay > 12 && actionHourOfDay <= 16) {
+      actionsByTimeOfDay.Afternoon += 1;
+    } else if (actionHourOfDay > 16 && actionHourOfDay <= 20) {
+      actionsByTimeOfDay.Evening += 1;
+    } else {
+      actionsByTimeOfDay.Night += 1;
+    }
+  });
+  return actionsByTimeOfDay;
+};
+
+// Takes object with {key: value} pairs of {timeOfDay: #-of-actions}
+// returns a date-sorted array in Recharts.js format
+export const formatActionsByTimeOfDay = (actionsByTimeOfDayObject) => {
+  const actionsByTimeOfDayArray = Object.entries(actionsByTimeOfDayObject);
+  return actionsByTimeOfDayArray.map((entry) => {
+    return {
+      timeOfDay: entry[0],
+      count: entry[1],
+    };
+  });
+};
+
+// Takes array of action objects and returns an object with {key: value} pairs of {verb: %-of-actions}
+export const getActionsByVerb = (actions) => {
+  const totalActions = actions.length;
+  const actionsByVerb = {};
+  actions.forEach((action) => {
+    if (!actionsByVerb[action.verb]) {
+      actionsByVerb[action.verb] = 1 / totalActions;
+    } else {
+      actionsByVerb[action.verb] += 1 / totalActions;
+    }
+  });
+  return actionsByVerb;
+};
+
+export const formatActionsByVerb = (actionsByVerbObject) => {
+  const actionsByVerbArray = Object.entries(actionsByVerbObject);
+
+  // capitalize verbs (entry[0][0]), convert 0.0x notation to x% and round to two decimal places (entry[0][1])
+  const formattedActionsByVerbArray = actionsByVerbArray
+    .map((entry) => [
+      entry[0][0].toUpperCase() + entry[0].substring(1),
+      parseFloat((entry[1] * 100).toFixed(2)),
+    ])
+    .filter((entry) => entry[1] >= MIN_PERCENTAGE_TO_SHOW_VERB);
+
+  // add ['other', x%] to cover all verbs that are filtered out of the array
+  if (formattedActionsByVerbArray.length) {
+    formattedActionsByVerbArray.push([
+      'Other',
```

Comment:
Can we use `t()` to translate here?

Reply:
I don't think so, since this is a CommonJS module, and in this case we would only be translating the verb 'Other' and not the others that are coming back from the server. I can look into translating verbs within the react component... but I think it's OK for now to keep them as is, since it is their server format. I was thinking however that it would be nice to add within the tooltips short explanations for what each verb means. in such a case, we would translate those explanations.

Reply:
OK, sounds good. Then factor this out as `OTHER_VERB`, or something like that. Note that I think that you access the `t` object from the i18n config, though. But it's true that it doesn't make sense to translate if we're not translating the verbs coming from the server.